scala/src/main/org/apache/spark/streaming/api/csharp/CSharpDStream.scala

```diff
@@ -343,3 +347,149 @@ class CSharpConstantInputDStream(ssc_ : StreamingContext, rdd: RDD[Array[Byte]])
 
   val asJavaDStream: JavaDStream[Array[Byte]] = JavaDStream.fromDStream(this)
 }
+
+case class RddPreComputeRecord[T] (
+    rddSeqNum: Long,
+    userInfo: Any,
+    rdd: RDD[T])
+
+/*
+ * Slurp RDDs, apply cache and pre-compute operation asynchronously, then spit them out
+ */
+class RddPreComputeCache[T](
+  sc: SparkContext,
+  val id: String,
+  val maxUnackedNum: Int,
+  val spitLimit: Int,       // maximum number of records for each spit
+  val numThreads: Int,
+  val storageLevel: StorageLevel) extends Logging {
+  // Thread pool is used to pre-compute RDDs. SeqNum is assigned to RDDs at input, so we can
+  // reassemble them at output
+  @volatile private var inputRddSeqNum: Long = 0L
+  @volatile private var outputRddSeqNum: Long = 0L
+
+  // ackedSeqNum is used for backpressure
+  @volatile private var ackedSeqNum: Long = -1L
+
+  private val precomputedRddMap: ConcurrentHashMap[Long, RddPreComputeRecord[T]] = new ConcurrentHashMap()
+
+  // If rdd is not pre-computed because of backpressure, it is put in this map
+  private val bypassRddMap: ConcurrentHashMap[Long, RddPreComputeRecord[T]] = new ConcurrentHashMap()
+
+  // If RDD is not pre-computed because of backpressure, it is put in this queue for future process
+  private val pendingRddQueue: LinkedBlockingQueue[RddPreComputeRecord[T]] =
+    new LinkedBlockingQueue[RddPreComputeRecord[T]]()
+
+  private var jobExecutor = ThreadUtils.newDaemonFixedThreadPool(numThreads, s"job-executor-$id")
+  private val pendingQueueScheduler = ThreadUtils.newDaemonSingleThreadExecutor(
+    s"pending-scheduler-$id")
+
+  private def isPending(rddSeqNum: Long): Boolean = {
+    rddSeqNum > ackedSeqNum + maxUnackedNum
+  }
+
+  private def processPendingQueue(): Unit = {
+    while (true) {
+      val record = pendingRddQueue.take()
+      logInfo(s"process pending queue [$id], rddSeqNum: ${record.rddSeqNum}, " +
+        s"ackedSeqNum: $ackedSeqNum, outputRddSeqNum: $outputRddSeqNum")
+      while (isPending(record.rddSeqNum)) {
+        Thread.sleep(100)
+      }
+      logInfo(s"process pending queue [$id], submitting job, rddSeqNum: ${record.rddSeqNum}, " +
+        s"ackedSeqNum: $ackedSeqNum, outputRddSeqNum: $outputRddSeqNum")
+      jobExecutor.execute(new Runnable {
```

Comment:
Consider extracting to a method and reuse between this usage and the usage in slurp()